pkg/agent/envoy_metrics_server.go

```diff
@@ -0,0 +1,57 @@
+package agent
+
+import (
+	"context"
+	envoyMetrics "github.com/datawire/ambassador/v2/pkg/api/envoy/service/metrics/v3"
+	"github.com/datawire/dlib/dlog"
+	"google.golang.org/grpc"
+	"io"
+	"net"
+)
+
+type streamHandler func(logCtx context.Context, in *envoyMetrics.StreamMetricsMessage)
+
+type metricsServer struct {
+	envoyMetrics.MetricsServiceServer
+	handler streamHandler
+	logCtx  context.Context
+}
+
+// NewMetricsServer is the main metricsServer constructor.
+func NewMetricsServer(handler streamHandler) *metricsServer {
+	return &metricsServer{
+		handler: handler,
+	}
+}
+
+// StartServer will start the metrics gRPC server, listening on :8006
+// It is a blocking call until grpcServer.Serve returns.
+func (s *metricsServer) StartServer(ctx context.Context) error {
+	grpcServer := grpc.NewServer()
+	envoyMetrics.RegisterMetricsServiceServer(grpcServer, s)
+
+	listener, err := net.Listen("tcp", ":8006")
+	if err != nil {
+		dlog.Errorf(ctx, "metrics service failed to listen: %v", err)
+	}
+
+	dlog.Infof(ctx, "metrics service listening on %s", listener.Addr().String())
+	s.logCtx = ctx
+	return grpcServer.Serve(listener)
```

Comment:
This should use github.com/datawire/dlib/dhttp, rather than the google.golang.org/grpc HTTP server. (For an example of how to do this, see the `cmd/example-envoy-metrics-sink/` which this PR also edits.)